src/js/core/fs/contract/IpfsInternalContract.ts

```diff
@@ -0,0 +1,134 @@
+import IInternalContract from "../../../models/interfaces/IInternalContract";
+import CallMessage from "../../rvm/lib/CallMessage";
+import { Results } from "../../rvm/context";
+import { toHex } from "../../rvm/utils/hexUtils";
+import { hexStringToString } from '../../../utils/hexUtils';
+
+import { getById, createOrUpdate} from '../../../services/DatabaseService'
+
+import BNType from 'bn.js';
+import { VM_ERROR } from "../../rvm/lib/exceptions";
+import Transaction from "../../../models/Transaction";
+const BN = require('bn.js');
+
+const MINIMAL_FILEHOST_DEPOSIT: BNType = new BN(1);
+const MINIMAL_DEPOSIT: BNType = new BN(0.001);
+
+/**
+ * Internal contract for the IPFS implementation
+ *
+ * data parameters
+ *
+ * [0 - 4] Always equals the selector of the function
+ *
+ * 0x00 00 00 01
+ * Deposit to file
+ * Params:
+ * [5 - 25] Address of the register
+ * value: Minimal of MINIMAL_DEPOSIT
+ * 
+ * 0x00 00 00 02
+ * Register as file host
+ * Params:
+ * [5 - 25] Address of the register
+ * value: Minimal of MINIMAL_DEPOSIT
+ *
+ * @class IpfsInteralContract
+ * @implements {IInternalContract}
+ */
+class IpfsInteralContract implements IInternalContract {
+    callMessage: CallMessage;
+    transaction: Transaction;
+    results: Results = {
+        exception: 0,
+        exceptionError: null,
+        gasUsed: 0,
+        outputRoot: '0x',
+        return: new Uint8Array(),
+        returnHex: '0x',
+    };
+
+    async depositToFile(): Promise<Results> {
+        if (this.callMessage.value.lt(MINIMAL_DEPOSIT)) {
+            this.results.exceptionError = VM_ERROR.REVERT;
+            return this.results;
+        }
+
+        // Read the last part of the input data and get the ipfs hash
+        const ipfsHashBytes = this.callMessage.inputData.slice(4, 51);
+        const ipfsHash = hexStringToString('0x' + toHex(ipfsHashBytes));
+
+        var ipfsFileObject = {
+            hash: ipfsHash,
+            value: 0,
+            hosts: [] as any
+        };
+
+        const dbIpfsFile = await getById(ipfsHash);
+        if(dbIpfsFile != null){
+            ipfsFileObject = dbIpfsFile;
+        }
+
+        // Add the stored Rutile to the file
+        ipfsFileObject.value += this.callMessage.value.toNumber();        
```

Comment:
toNumber() could be "dangerous". Because values can be higher than 32-bit. Use the BN adding instead.